### DISCUSSION TITLE
Unit-testing autodispose providers

### DISCUSSION THREAD

**Discussion by derolf** [rrousselGit/riverpod#2503] [Q&A] [answered]
I am trying to unittest some code involving autodispose providers. Dart only, no Flutter widgets.

However, whenever I pump the ProviderContainer, they get disposed.

What’s a good approach to keep them alive?

**Comment by rrousselGit** [comment]
Don't "read" them. Use container.listen or have another provider watch them.

**Comment by derolf** [comment, marked as answer]
I found out that I can get a KeepAliveLink, which works pretty well.

**Reply by ruchitmavani** [reply]
Hey having same issue, can you provide example how you achieve it?

**Comment by derolf** [comment]
A more detailed answer:

```dart

// if provider implementation has "ref" field
final keepAlive = container.read(someNotifierProvider.notifier).ref.keepAlive();

// this did not work for me:
// final keepAlive = container.listen(someNotifierProvider.notifier, (_, __) {});

...

keepAlive.close();

**Reply by rrousselGit** [reply]
Don't do that. Use container.listen line I told you,

**Reply by rrousselGit** [reply]
```dart
final sub = container.listen(provider, (_, __) {});

expect(sub.read(), value);
await something;
expect(sub.read(), another);
```

**Reply by derolf** [reply]
`container.listen` did not work for me for some reason.

**Reply by rrousselGit** [reply]
It should. That's how all autoDispose providers are tested within Riverpod.
